File: implementations/go/examples/internal/infrastructure/persistence/account_repository.go
package persistence

import (
	"context"
	"database/sql"
	"fmt"
	"strings"
	"time"

	"github.com/example/account-service/internal/domain/account"
	"github.com/example/account-service/internal/infrastructure/database"
	"github.com/example/account-service/internal/infrastructure/outbox"
)

type AccountRepository struct {
	db           *sql.DB
	outboxWriter *outbox.Writer
}

// Compile-time interface satisfaction check — AccountRepository satisfies
// both Repository (Command) and Query. Since Go interfaces use structural
// typing, the same concrete struct serves both roles without needing two
// separate implementations.
var _ account.Repository = (*AccountRepository)(nil)
var _ account.Query = (*AccountRepository)(nil)

func NewAccountRepository(db *sql.DB, outboxWriter *outbox.Writer) *AccountRepository {
	return &AccountRepository{db: db, outboxWriter: outboxWriter}
}

func (r *AccountRepository) FindAccounts(ctx context.Context, q account.FindQuery) ([]*account.Account, int, error) {
	args := []any{}
	where := []string{"deleted_at IS NULL"}
	i := 1

	if q.AccountID != "" {
		where = append(where, fmt.Sprintf("id = $%d", i))
		args = append(args, q.AccountID)
		i++
	}
	if q.OwnerID != "" {
		where = append(where, fmt.Sprintf("owner_id = $%d", i))
		args = append(args, q.OwnerID)
		i++
	}
	if len(q.Status) > 0 {
		placeholders := make([]string, len(q.Status))
		for j, s := range q.Status {
			placeholders[j] = fmt.Sprintf("$%d", i)
			args = append(args, string(s))
			i++
		}
		where = append(where, fmt.Sprintf("status IN (%s)", strings.Join(placeholders, ",")))
	}

	whereClause := strings.Join(where, " AND ")

	var total int
	if err := r.db.QueryRowContext(ctx,
		fmt.Sprintf(`SELECT COUNT(*) FROM accounts WHERE %s`, whereClause), args...,
	).Scan(&total); err != nil {
		return nil, 0, fmt.Errorf("count accounts: %w", err)
	}

	args = append(args, q.Take, q.Page*q.Take)
	rows, err := r.db.QueryContext(ctx,
		fmt.Sprintf(`SELECT id, owner_id, email, amount, currency, status, created_at, updated_at, last_interest_paid_at
		 FROM accounts WHERE %s ORDER BY id DESC LIMIT $%d OFFSET $%d`, whereClause, i, i+1),
		args...,
	)
	if err != nil {
		return nil, 0, fmt.Errorf("find accounts: %w", err)
	}
	defer func() { _ = rows.Close() }()

	var accounts []*account.Account
	for rows.Next() {
		var id, ownerID, email, currency, status string
		var amount int64
		var createdAt, updatedAt time.Time
		var lastInterestPaidAt sql.NullTime
		if err := rows.Scan(&id, &ownerID, &email, &amount, &currency, &status, &createdAt, &updatedAt, &lastInterestPaidAt); err != nil {
			return nil, 0, err
		}
		balance, err := account.NewMoney(amount, currency)
		if err != nil {
			return nil, 0, err
		}
		accounts = append(accounts, account.Reconstitute(id, ownerID, email, balance, account.Status(status), createdAt, updatedAt, lastInterestPaidAt.Time))
	}
	return accounts, total, rows.Err()
}

// SaveAccount, if there's an ambient transaction (the *sql.Tx that
// database.WithTx/Manager.RunInTx stashed in ctx), only participates in it
// and leaves committing to the caller (TransferHandler, etc.) — if there
// isn't one (same as every existing standalone caller), it opens its own
// transaction and is responsible for committing it too. Both paths share
// the same underlying SQL via saveAccount, so the behavior of existing
// callers with no ambient transaction doesn't change in the slightest.
func (r *AccountRepository) SaveAccount(ctx context.Context, a *account.Account) error {
	if tx, ok := database.TxFromContext(ctx); ok {
		if err := r.saveAccount(ctx, tx, a); err != nil {
			return err
		}
		// Committing is the ambient transaction owner's job (TransferHandler,
		// etc.), so we can't actually confirm commit success here — but this
		// is harmless because the new use case (Transfer) that takes this
		// path never reuses or retries the same *Account instance after this
		// call.
		a.ClearTransactions()
		a.ClearEvents()
		return nil
	}

	tx, err := r.db.BeginTx(ctx, nil)
	if err != nil {
		return fmt.Errorf("begin tx: %w", err)
	}
	defer func() { _ = tx.Rollback() }()

	if err := r.saveAccount(ctx, tx, a); err != nil {
		return err
	}
	if err := tx.Commit(); err != nil {
		return fmt.Errorf("commit save account: %w", err)
	}
	// Same as the existing behavior — the in-memory pending state is only
	// cleared after actually confirming commit success. On commit failure, a
	// retry must be able to try the transaction/event again.
	a.ClearTransactions()
	a.ClearEvents()
	return nil
}

// saveAccount runs 3 statements — saving the account, saving the
// transaction, and loading the Outbox — using the given tx. Committing/
// rolling back is the caller's responsibility (SaveAccount or the ambient
// transaction owner).
func (r *AccountRepository) saveAccount(ctx context.Context, tx *sql.Tx, a *account.Account) error {
	var lastInterestPaidAt any
	if !a.LastInterestPaidAt.IsZero() {
		lastInterestPaidAt = a.LastInterestPaidAt
	}
	_, err := tx.ExecContext(ctx,
		`INSERT INTO accounts (id, owner_id, email, amount, currency, status, updated_at, last_interest_paid_at)
		 VALUES ($1, $2, $3, $4, $5, $6, NOW(), $7)
		 ON CONFLICT (id) DO UPDATE SET amount = EXCLUDED.amount, status = EXCLUDED.status,
		   updated_at = NOW(), last_interest_paid_at = EXCLUDED.last_interest_paid_at`,
		a.AccountID, a.OwnerID, a.Email, a.Balance.Amount, a.Balance.Currency, string(a.Status), lastInterestPaidAt,
	)
	if err != nil {
		return fmt.Errorf("save account: %w", err)
	}

	for _, t := range a.PendingTransactions() {
		var referenceID any
		if t.ReferenceID != "" {
			referenceID = t.ReferenceID
		}
		_, err = tx.ExecContext(ctx,
			`INSERT INTO transactions (id, account_id, type, amount, currency, reference_id, created_at)
			 VALUES ($1, $2, $3, $4, $5, $6, $7)`,
			t.TransactionID, t.AccountID, string(t.Type), t.Amount.Amount, t.Amount.Currency, referenceID, t.CreatedAt,
		)
		if err != nil {
			return fmt.Errorf("save transaction: %w", err)
		}
	}

	// The Outbox row is loaded within the same transaction as the account/
	// transaction rows — since the commit is atomic, the "account changed
	// but the event was lost" (dual-write) failure mode cannot occur.
	if err := r.outboxWriter.SaveAll(ctx, tx, a.DomainEvents()); err != nil {
		return err
	}

	return nil
}

// HasTransactionWithReference implements the idempotency check required by
// account.Query. See the account.Query interface's comment for why it must
// be checked as the (referenceID, type) combination.
func (r *AccountRepository) HasTransactionWithReference(ctx context.Context, referenceID string, txType account.TransactionType) (bool, error) {
	var count int
	if err := r.db.QueryRowContext(ctx,
		`SELECT COUNT(*) FROM transactions WHERE reference_id = $1 AND type = $2`,
		referenceID, string(txType),
	).Scan(&count); err != nil {
		return false, fmt.Errorf("count transactions by reference: %w", err)
	}
	return count > 0, nil
}

// FindTransactions applies q's optional Type/FromDate/ToDate filters as
// additional WHERE clauses on top of the mandatory account_id match — the
// same dynamic-predicate-building idiom PaymentRepository.FindPayments uses
// for its CreatedFrom/CreatedTo range. FromDate/ToDate arrive as plain ISO
// 8601 date strings (see FindTransactionsQuery's doc comment); a value that
// fails to parse is treated as "no bound on that side" rather than an
// error, so a malformed filter (e.g. from an LLM Technical Service that
// already validates this itself, but is not the only possible caller)
// degrades to "no narrowing" instead of blocking the query.
func (r *AccountRepository) FindTransactions(ctx context.Context, q account.FindTransactionsQuery) ([]account.Transaction, int, error) {
	args := []any{q.AccountID}
	where := []string{"account_id = $1"}
	i := 2

	if q.Type != "" {
		where = append(where, fmt.Sprintf("type = $%d", i))
		args = append(args, string(q.Type))
		i++
	}
	if from, ok := parseISODate(q.FromDate); ok {
		where = append(where, fmt.Sprintf("created_at >= $%d", i))
		args = append(args, from)
		i++
	}
	if to, ok := parseISODate(q.ToDate); ok {
		// Exclusive next-day-start bound makes ToDate an inclusive end-of-day
		// boundary without needing a separate "23:59:59.999" literal.
		where = append(where, fmt.Sprintf("created_at < $%d", i))
		args = append(args, to.AddDate(0, 0, 1))
		i++
	}
	whereClause := strings.Join(where, " AND ")

	var total int
	if err := r.db.QueryRowContext(ctx,
		fmt.Sprintf(`SELECT COUNT(*) FROM transactions WHERE %s`, whereClause), args...,
	).Scan(&total); err != nil {
		return nil, 0, fmt.Errorf("count transactions: %w", err)
	}

	limitArgs := append(append([]any{}, args...), q.Take, q.Page*q.Take)
	rows, err := r.db.QueryContext(ctx,
		fmt.Sprintf(`SELECT id, account_id, type, amount, currency, created_at
		 FROM transactions WHERE %s
		 ORDER BY created_at DESC LIMIT $%d OFFSET $%d`, whereClause, i, i+1),
		limitArgs...,
	)
	if err != nil {
		return nil, 0, fmt.Errorf("find transactions: %w", err)
	}
	defer func() { _ = rows.Close() }()

	var transactions []account.Transaction
	for rows.Next() {
		var t account.Transaction
		var txType, currency string
		var amount int64
		if err := rows.Scan(&t.TransactionID, &t.AccountID, &txType, &amount, &currency, &t.CreatedAt); err != nil {
			return nil, 0, err
		}
		t.Type = account.TransactionType(txType)
		t.Amount = account.Money{Amount: amount, Currency: currency}
		transactions = append(transactions, t)
	}
	return transactions, total, rows.Err()
}

// SummarizeTransactions aggregates one account's transactions in a
// [CreatedFrom, CreatedTo) date range, filtered by type — the same
// dynamic-predicate-building idiom as FindTransactions, but returning only
// the count/total (COUNT(*)/COALESCE(SUM(...), 0)) rather than the
// individual rows, used by AnalyzeMonthlySpendingHandler to total a month's
// (and the prior month's) WITHDRAWAL activity without loading every
// Transaction row into memory.
func (r *AccountRepository) SummarizeTransactions(ctx context.Context, q account.SummarizeTransactionsQuery) (account.TransactionSummary, error) {
	args := []any{q.AccountID}
	where := []string{"account_id = $1"}
	i := 2

	if len(q.Type) > 0 {
		placeholders := make([]string, len(q.Type))
		for j, t := range q.Type {
			placeholders[j] = fmt.Sprintf("$%d", i)
			args = append(args, string(t))
			i++
		}
		where = append(where, fmt.Sprintf("type IN (%s)", strings.Join(placeholders, ",")))
	}
	where = append(where, fmt.Sprintf("created_at >= $%d", i))
	args = append(args, q.CreatedFrom)
	i++
	where = append(where, fmt.Sprintf("created_at < $%d", i))
	args = append(args, q.CreatedTo)

	whereClause := strings.Join(where, " AND ")

	var summary account.TransactionSummary
	if err := r.db.QueryRowContext(ctx,
		fmt.Sprintf(`SELECT COUNT(*), COALESCE(SUM(amount), 0) FROM transactions WHERE %s`, whereClause), args...,
	).Scan(&summary.Count, &summary.TotalAmount); err != nil {
		return account.TransactionSummary{}, fmt.Errorf("summarize transactions: %w", err)
	}
	return summary, nil
}

// parseISODate parses an ISO 8601 date (YYYY-MM-DD). An empty or
// unparsable value returns ok=false so the caller can skip applying that
// bound rather than erroring.
func parseISODate(v string) (time.Time, bool) {
	if v == "" {
		return time.Time{}, false
	}
	t, err := time.Parse("2006-01-02", v)
	if err != nil {
		return time.Time{}, false
	}
	return t, true
}

File: implementations/fastapi/examples/src/account/infrastructure/persistence/account_repository.py
from datetime import date, datetime, time

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

from ....common.clock import utc_now
from ...domain.account import Account
from ...domain.account_status import AccountStatus
from ...domain.money import Money
from ...domain.repository import AccountRepository
from ...domain.transaction import Transaction, TransactionType


class Base(DeclarativeBase):
    pass


class AccountModel(Base):
    __tablename__ = "accounts"

    id: Mapped[str] = mapped_column(primary_key=True)
    owner_id: Mapped[str]
    email: Mapped[str]
    amount: Mapped[int]
    currency: Mapped[str]
    status: Mapped[str]
    created_at: Mapped[datetime] = mapped_column(default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(default=utc_now, onupdate=utc_now)
    deleted_at: Mapped[datetime | None] = mapped_column(nullable=True, default=None)
    # The Level 1 idempotency marker for the regular interest-payment batch — see Account.apply_interest().
    last_interest_paid_at: Mapped[date | None] = mapped_column(nullable=True, default=None)


class TransactionModel(Base):
    __tablename__ = "transactions"

    id: Mapped[str] = mapped_column(primary_key=True)
    account_id: Mapped[str]
    type: Mapped[str]
    amount: Mapped[int]
    currency: Mapped[str]
    created_at: Mapped[datetime] = mapped_column(default=utc_now)
    # Correlates a transaction left by a reaction to the Payment BC's Integration Event
    # (payment_id/refund_id), and also serves as the Level 2 Ledger key that prevents
    # reprocessing the same (reference_id, type) combination (see
    # has_transaction_with_reference). Absent (nullable) for a deposit/withdrawal the user
    # requested directly.
    reference_id: Mapped[str | None] = mapped_column(nullable=True, index=True, default=None)
    # The payee/memo optionally attached to a withdrawal at request time — see transaction.py.
    merchant_name: Mapped[str | None] = mapped_column(nullable=True, default=None)
    # Filled in asynchronously by CategorizeTransactionEventHandler — null until that
    # reaction runs (or forever, for a transaction with no merchant_name to classify).
    category: Mapped[str | None] = mapped_column(nullable=True, default=None)


class SqlAlchemyAccountRepository(AccountRepository):
    def __init__(self, session: AsyncSession) -> None:
        # deferred import — since outbox_model.py imports this module's Base, importing
        # OutboxWriter at the module's top level would create a circular import (see
        # "Python's circular imports" in module-pattern.md).
        from ....outbox.outbox_writer import OutboxWriter

        self._session = session
        self._outbox_writer = OutboxWriter(session)

    async def find_accounts(
        self,
        page: int,
        take: int,
        account_id: str | None = None,
        owner_id: str | None = None,
        status: list[str] | None = None,
    ) -> tuple[list[Account], int]:
        stmt = select(AccountModel).where(AccountModel.deleted_at.is_(None))
        count_stmt = select(func.count()).select_from(AccountModel).where(AccountModel.deleted_at.is_(None))

        if account_id:
            stmt = stmt.where(AccountModel.id == account_id)
            count_stmt = count_stmt.where(AccountModel.id == account_id)
        if owner_id:
            stmt = stmt.where(AccountModel.owner_id == owner_id)
            count_stmt = count_stmt.where(AccountModel.owner_id == owner_id)
        if status:
            stmt = stmt.where(AccountModel.status.in_(status))
            count_stmt = count_stmt.where(AccountModel.status.in_(status))

        total = (await self._session.execute(count_stmt)).scalar_one()
        rows = (
            (await self._session.execute(stmt.order_by(AccountModel.id.desc()).offset(page * take).limit(take)))
            .scalars()
            .all()
        )

        return [self._to_domain(row) for row in rows], total

    async def save_account(self, account: Account) -> None:
        existing = await self._session.get(AccountModel, account.account_id)
        if existing:
            existing.amount = account.balance.amount
            existing.status = account.status.value
            existing.last_interest_paid_at = account.last_interest_paid_at
            existing.updated_at = utc_now()
        else:
            self._session.add(
                AccountModel(
                    id=account.account_id,
                    owner_id=account.owner_id,
                    email=account.email,
                    amount=account.balance.amount,
                    currency=account.balance.currency,
                    status=account.status.value,
                    last_interest_paid_at=account.last_interest_paid_at,
                )
            )

        for transaction in account.pull_pending_transactions():
            self._session.add(
                TransactionModel(
                    id=transaction.transaction_id,
                    account_id=transaction.account_id,
                    type=transaction.type,
                    amount=transaction.amount.amount,
                    currency=transaction.amount.currency,
                    created_at=transaction.created_at,
                    reference_id=transaction.reference_id,
                    merchant_name=transaction.merchant_name,
                )
            )

        events = account.pull_events()
        if events:
            await self._outbox_writer.save_all(events)

        await self._session.flush()

    async def find_transactions(
        self,
        account_id: str,
        page: int,
        take: int,
        type: TransactionType | None = None,
        from_date: date | None = None,
        to_date: date | None = None,
    ) -> tuple[list[Transaction], int]:
        stmt = select(TransactionModel).where(TransactionModel.account_id == account_id)
        count_stmt = select(func.count()).select_from(TransactionModel).where(TransactionModel.account_id == account_id)

        if type:
            stmt = stmt.where(TransactionModel.type == type)
            count_stmt = count_stmt.where(TransactionModel.type == type)
        if from_date:
            from_dt = datetime.combine(from_date, time.min)
            stmt = stmt.where(TransactionModel.created_at >= from_dt)
            count_stmt = count_stmt.where(TransactionModel.created_at >= from_dt)
        if to_date:
            to_dt = datetime.combine(to_date, time.max)
            stmt = stmt.where(TransactionModel.created_at <= to_dt)
            count_stmt = count_stmt.where(TransactionModel.created_at <= to_dt)

        total = (await self._session.execute(count_stmt)).scalar_one()
        rows = (
            (
                await self._session.execute(
                    stmt.order_by(TransactionModel.created_at.desc()).offset(page * take).limit(take)
                )
            )
            .scalars()
            .all()
        )

        transactions = [
            Transaction(
                transaction_id=r.id,
                account_id=r.account_id,
                type=r.type,  # type: ignore[arg-type]
                amount=Money(r.amount, r.currency),
                created_at=r.created_at,
                reference_id=r.reference_id,
                merchant_name=r.merchant_name,
                category=r.category,  # type: ignore[arg-type]
            )
            for r in rows
        ]
        return transactions, total

    async def has_transaction_with_reference(self, reference_id: str, type: str) -> bool:
        stmt = (
            select(func.count())
            .select_from(TransactionModel)
            .where(TransactionModel.reference_id == reference_id, TransactionModel.type == type)
        )
        count = (await self._session.execute(stmt)).scalar_one()
        return count > 0

    async def summarize_transactions(
        self,
        account_id: str,
        type: list[TransactionType],
        created_at_from: datetime,
        created_at_to: datetime,
    ) -> tuple[int, int]:
        stmt = (
            select(func.count(), func.coalesce(func.sum(TransactionModel.amount), 0))
            .select_from(TransactionModel)
            .where(
                TransactionModel.account_id == account_id,
                TransactionModel.type.in_(type),
                TransactionModel.created_at >= created_at_from,
                TransactionModel.created_at < created_at_to,
            )
        )
        count, total_amount = (await self._session.execute(stmt)).one()
        return int(count), int(total_amount)

    def _to_domain(self, row: AccountModel) -> Account:
        return Account(
            account_id=row.id,
            owner_id=row.owner_id,
            email=row.email,
            balance=Money(row.amount, row.currency),
            status=AccountStatus(row.status),
            created_at=row.created_at,
            updated_at=row.updated_at,
            last_interest_paid_at=row.last_interest_paid_at,
        )
